### ISSUE TITLE
Chebyshev Propagation Time Evolution

### ISSUE THREAD

**Issue by refraction-ray** [tensorcircuit/tensorcircuit-ng#23]
# TEP - Chebyshev Propagation Time Evolution

**Author**
@refraction-ray

**Status**
Draft

**Created**
<2025-07-28>

## Abstract

This TEP proposes the addition of a Chebyshev polynomial expansion method for simulating the time evolution of quantum states, `ψ(t) = exp(-iHt)ψ(0)`. This method is particularly renowned for its high precision and low memory footprint in long-time simulations of large quantum systems. The proposed function, `chebyshev_evol`, will be implemented to be fully compatible with TensorCircuit's machine learning backend, supporting automatic differentiation (AD), vectorized mapping (vmap), just-in-time (JIT) compilation, and GPU acceleration. It will offer an alternative to exact exponentiation and Krylov methods, excelling in scenarios requiring fixed-time, high-accuracy evolution.

## Motivation and Scope

While exact evolution via `hamiltonian_evol` is limited to small systems and Krylov methods (proposed in a separate TEP： https://github.com/tensorcircuit/tensorcircuit-ng/issues/13) are ideal for flexible time-stepping over short durations, there is a clear need for a solver optimized for **long-time, high-precision, memory-efficient** simulations. This is a common requirement in studying quantum chaos, thermalization, and many-body localization, where evolution over thousands of natural time units is necessary.

The Chebyshev propagation method addresses this need. It approximates the evolution operator `exp(-iHt)` using a global expansion in Chebyshev polynomials. The key advantages are:
1.  **Numerical Stability and Accuracy**: The error of the approximation is near-uniformly distributed over the entire time interval and decreases exponentially with the number of expansion terms.
2.  **Low Memory Usage**: The computation relies on a three-term recurrence relation, requiring only a few state vectors to be stored in memory at any given time, making it ideal for extremely large Hilbert spaces.
3.  **Efficiency for Fixed-Time Evolution**: For a given final time `t`, the number of terms and thus the computational cost is fixed and predictable.

The primary prerequisite is knowledge of the Hamiltonian's spectral bounds (`E_min`, `E_max`), which are used to normalize the Hamiltonian spectrum to the `[-1, 1]` interval where Chebyshev polynomials are defined.

**Scope:**
- Implement a function `chebyshev_evol` that evolves a state `psi0` to a final time `t_final`.
- The core algorithm will be based on the Chebyshev polynomial expansion of the `exp(-ix)` function, using a three-term recurrence to apply the polynomial operator to the state vector.
- The implementation must leverage TensorCircuit's backend primitives to be compatible with AD, vmap, JIT, and GPU execution.
- It must handle the mandatory spectral normalization of the Hamiltonian. An auxiliary function to estimate the spectral bounds (e.g., using Lanczos) will be a valuable addition.

## Usage and Impact

```python
# Estimate spectral bounds (E_max, E_min) first, e.g., using a Lanczos utility
E_max, E_min = tc.experimental.estimate_spectral_bounds(h, n_iter=30)

# Evolve to a single final time t_final
psi_final = tc.experimental.chebyshev_evol(
    h=h,
    psi0=psi0,
    t_final=100.0,
    spectral_bounds=(E_max, E_min),
    n_terms=500  # Number of terms determines accuracy
)

# Optional: evolve to a list of times and get intermediate states
# (Note: this is less efficient than Krylov for this task)
results = tc.experimental.chebyshev_evol(
    h=h,
    psi0=psi0,
    tlist=np.linspace(0, 100, 11),
    spectral_bounds=(E_max, E_min),
    n_terms=500,
    callback=compute_expectation
)
```

**Impact:** This addition will significantly enhance TensorCircuit's capabilities as a high-performance numerical solver for many-body physics, particularly for long-time dynamics in Rydberg atom arrays and other analog simulation platforms. It complements the existing and proposed solvers by filling the niche for memory-efficient, high-precision, fixed-time evolution.

## Backward compatibility

No issue. This is a new, additive feature.

## Related Work

- **SciPy**: `scipy.sparse.linalg.expm_multiply` uses a Krylov method, but Chebyshev is a well-known alternative.

## Implementation

1.  **Spectral Normalization:**
    - The `chebyshev_evol` function will take `spectral_bounds=(E_max, E_min)` as a mandatory argument.
    - Internally, it will define a normalized Hamiltonian `H_norm = (2*H - (E_max + E_min)*I) / (E_max - E_min)`, whose spectrum lies in `[-1, 1]`.
    - The evolution is then rewritten in terms of `H_norm` and rescaled time.
2.  **Core Chebyshev Recurrence:**
    - Implement the iterative loop `|T_{k+1}⟩ = 2*H_norm|T_k⟩ - |T_{k-1}⟩`, where `|T_k⟩ = T_k(H_norm)|psi0⟩`.
    - This loop must be implemented using backend-compatible scan operations (`jax.lax.scan`, `tf.scan`) to be JIT-compilable.
    - The loop computes the sequence of vectors `{|T_k⟩}`.
3.  **State Reconstruction:**
    - The final state is a weighted sum: `ψ(t) = Σ_k c_k(t) |T_k⟩`.
    - The coefficients `c_k(t) = (2 - δ_{k,0}) * (-i)^k * J_k(τ)`, where `J_k` is the k-th Bessel function of the first kind and `τ` is the rescaled time.
    - Pre-compute the Bessel function values using the backend's `scipy.special` bindings (e.g., `jax.scipy.special.jv`).
    - The summation can be performed efficiently within or after the scan loop.
4.  **Function Interface:**
    - Create the `chebyshev_evol(h, psi0, t_final, spectral_bounds, n_terms, tlist=None, callback=None)` function signature.
    - If `tlist` is provided, the function will compute the state at each time point in the list. This will involve re-calculating the coefficients `c_k(t)` for each `t` but reusing the same set of vectors `{|T_k⟩}`.
    - Include input validation (e.g., `E_max > E_min`, `n_terms > 0`).
5.  **Utility for Spectral Bounds:**
    - Implement a helper function `estimate_spectral_bounds(h, n_iter=30)` that uses a few iterations of the Lanczos algorithm to quickly and reliably estimate the largest and smallest eigenvalues of `h`. This is a crucial utility for user convenience.
6.  **Testing:**
    - **Accuracy Tests**: Compare results against `hamiltonian_evol` for small systems. Verify that the error decreases exponentially as `n_terms` increases.
    - **Long-time Stability Tests**: Evolve for a long time and verify norm conservation, comparing its stability against a naive multi-step Krylov approach.
    - **Memory Tests**: Profile memory usage and confirm it scales with `D` (Hilbert space size) and not with `n_terms` or `t_final`.
    - **Backend, AD, vmap, JIT, GPU Tests**: As detailed in the Krylov TEP, ensure full compatibility with the TensorCircuit ecosystem.

## Alternatives

- **Real-Time Evolution with Polynomials (RETEP)**: A similar polynomial expansion method, but can sometimes be less stable than Chebyshev for very long times.
- **Taylor Series Expansion**: Numerically unstable for large times and generally not used for this purpose.
- **Krylov Subspace Methods**: As discussed, they are an excellent alternative but optimized for a different use case (flexible, short-time evolution). Chebyshev propagation is superior for fixed, long-time, memory-constrained problems.

## References

* P. Sierant, E. G. Lazo, M. Dalmonte, A. Scardicchio, and J. Zakrzewski, Constraint-induced delocalization, Phys. Rev. Lett. 127, 126603 (2021).
